compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1518,7 +1518,15 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
 
     /** Replace each alternative by its apply members where necessary */
     def applyMembers(alt: TermRef): List[TermRef] =
-      if (tryApply(alt)) alt.member(nme.apply).alternatives.map(TermRef(alt, nme.apply, _))
+      if (tryApply(alt)) {
+        val qual = alt.widen match {
+          case pt: PolyType =>
+            pt.resultType
+          case _ =>
+            alt
+        }
+        qual.member(nme.apply).alternatives.map(TermRef(alt, nme.apply, _))
```

Comment:
I find this extremely dicey. alt.apply is not even a well formed term and its type refers an unbound param ref. So this does not look right to me. Scala 2's behavior is no excuse, since it has many situations where bound variables escape. So I do not think this is the right fix, and I am not sure we need a fix at all, since it seems impossible to describe the fix in terms that are simple and sound.

Reply:
> alt.apply is not even a well formed term

But this is also true for the case where `alt` is a monomorphic method since it's not a stable prefix, and yet the existing code will happily make up a TermRef for it. If we wanted to be stricter we could use `.narrow` instead.

> its type refers an unbound param ref.

True, since this is only used for overloading resolution, it seemed OK to me. If we want to avoid that, an alternative would be to make the resulting types polymorphic, e.g. if we start with:
```scala
def foo2[T]: T => String
```
Then we could pretend that it really has type:
```scala
[T](x: T): String
```

This would mean that switching between having a polymorphic method with a parameter list and a polymorphic method that returns an object with the same parameter list would not affect which overload gets chosen, this is a nice property to have and it matches what happens with monomorphic methods. The only downside I see is that it doesn't match Scala 2 behavior, but I don't think that's a big deal here (and the current behavior of dotty master can also select a different overload than Scala 2, so we wouldn't be regressing).

Reply:
You'd certainly get into problems if the polytype is already in use in the current constraint. Yes, we can move the PolyType around, onto the apply method, but these are complicated acrobatics which make the code base harder to understand. Do we have a hard reason for doing this? Being compatible with scalac is not reason enough by itself, I think.